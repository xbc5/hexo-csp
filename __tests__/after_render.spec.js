"use strict";
const { fixture } = require("./helpers/mock");
const path = require("path");

describe("given mixed mode policies and env=prod", () => {
  ["markdown-1", "markdown-2", "/"].forEach((name) => {
    describe(`for ${name}`, () => {
      it("should merge and render the given policies", async () => {
        const render = await fixture("mixed-mode-prod", { env: "prod" });
        expect(await render(name)).toMatchSnapshot();
      });
    });
  });
});

describe("given mixed mode policies and env=dev", () => {
  ["markdown-1", "markdown-2", "/"].forEach((name) => {
    describe(`for ${name}`, () => {
      it("should merge and render the given policies", async () => {
        const render = await fixture("mixed-mode-dev", { env: "dev" });
        expect(await render(name)).toMatchSnapshot();
      });
    });
  });
});

// TODO: test prod replace mode
describe("given replace mode policies and env=dev", () => {
  ["markdown-1", "markdown-2", "/"].forEach((name) => {
    describe(`for ${name}`, () => {
      it("should replace where appropriate and render the given policies", async () => {
        const render = await fixture("replace-mode-dev", { env: "dev" });
        expect(await render(name)).toMatchSnapshot();
      });
    });
  });
});

describe("given invalid frontmatter", () => {
  ["markdown-1", "/"].forEach((name) => {
    describe(`for ${name}`, () => {
      it("should not render a CSP", async () => {
        const render = await fixture("invalid-frontmatter");
        expect(await render(name)).toMatchSnapshot();
      });
    });
  });
});

describe("given the custom permalink", () => {
  ["2001/01/01/markdown-1", "/", "page-1/"].forEach((p) => {
    describe(p, () => {
      const basename = p === "/" ? "/" : path.basename(p);
      it(`should ignore the dirname and use only the basename: ${basename}`, async () => {
        const render = await fixture("custom-permalink", { env: "prod" });
        expect(await render(p)).toMatchSnapshot();
      });
    });
  });
});

["/", "/index.html", "index.html"].forEach((path) => {
  describe(`given the path "${path}"`, () => {
    it(`should only match against the "/" policy path`, async () => {
      const render = await fixture("slash-root", { env: "prod" });
      expect(await render(path)).toMatchSnapshot();
    });
  });
});

/*
WARN: snapshots are prettified by jest-serializer-html
don't rely on them to reflect how the document is truly rendered.
*/
describe("when inline tag sources are enabled", () => {
  ["/", "markdown-1", "markdown-2"].forEach((path) => {
    describe(`path: "${path}"`, () => {
      it(`should contain a hash+base64 source for each script/style, that merges with existing policies`, async () => {
        const render = await fixture("inline-tags", { env: "prod" });
        expect(await render(path)).toMatchSnapshot();
      });
    });
  });
});
