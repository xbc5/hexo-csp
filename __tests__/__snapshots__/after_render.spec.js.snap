// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given invalid frontmatter for / should not render a CSP 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>
      Markdown Home
    </title>
  </head>
</html>
<body>
  <h2 id="Home">
    Home
  </h2>
</body>
`;

exports[`given invalid frontmatter for markdown-1 should not render a CSP 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>
      Markdown 1
    </title>
  </head>
</html>
<body>
  <h1>
    Markdown 1
  </h1>
</body>
`;

exports[`given mixed mode policies and env=dev for / should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://frontmatter-home-default.com; img-src https://frontmatter-home-img.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h2 id="Home">
      Home
    </h2>
  </body>
</html>
`;

exports[`given mixed mode policies and env=dev for markdown-1 should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="img-src https://dev-markdown-1-img.com https://frontmatter-markdown-1-img-1.com https://frontmatter-markdown-1-img-2.com; default-src https://frontmatter-markdown-1-default-1.com https://frontmatter-markdown-1-default-2.com"
    >
    <title>
      Markdown 1
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given mixed mode policies and env=dev for markdown-2 should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://dev-markdown-2-default-1.com https://dev-markdown-2-default-2.com https://frontmatter-markdown-2-default-1.com https://frontmatter-markdown-2-default-2.com; img-src https://dev-markdown-2-img.com https://frontmatter-markdown-2-img-1.com https://frontmatter-markdown-2-img-2.com"
    >
    <title>
      Markdown 2
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given mixed mode policies and env=prod for / should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-frontmatter-home-default.com; img-src https://prod-frontmatter-home-img.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h2 id="Home">
      Home
    </h2>
  </body>
</html>
`;

exports[`given mixed mode policies and env=prod for markdown-1 should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-markdown-1-default-1.com https://prod-markdown-1-default-2.com https://frontmatter-markdown-1-default-1.com https://frontmatter-markdown-1-default-2.com; img-src https://prod-markdown-1-img.com https://frontmatter-markdown-1-img-1.com https://frontmatter-markdown-1-img-2.com"
    >
    <title>
      Markdown 1
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given mixed mode policies and env=prod for markdown-2 should merge and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="img-src https://prod-markdown-2-img.com https://frontmatter-markdown-2-img-1.com https://frontmatter-markdown-2-img-2.com; default-src https://frontmatter-markdown-2-default-1.com https://frontmatter-markdown-2-default-2.com"
    >
    <title>
      Markdown 2
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given replace mode policies and env=dev for / should replace where appropriate and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-config-index-default.com https://dev-frontmatter-home-default.com; img-src https://dev-frontmatter-home-img.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h2 id="Home">
      Home
    </h2>
  </body>
</html>
`;

exports[`given replace mode policies and env=dev for markdown-1 should replace where appropriate and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://dev-frontmatter-markdown-1-default-1.com https://dev-frontmatter-markdown-1-default-2.com; img-src https://dev-frontmatter-markdown-1-img-1.com https://dev-frontmatter-markdown-1-img-2.com"
    >
    <title>
      Markdown 1
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given replace mode policies and env=dev for markdown-2 should replace where appropriate and render the given policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://dev-frontmatter-markdown-2-default-1.com https://dev-frontmatter-markdown-2-default-2.com; img-src https://dev-frontmatter-markdown-2-img-1.com https://dev-frontmatter-markdown-2-img-2.com"
    >
    <title>
      Markdown 2
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given the custom permalink / should ignore the dirname and use only the basename: / 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-index-default.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h1>
      Home
    </h1>
  </body>
</html>
`;

exports[`given the custom permalink 2001/01/01/markdown-1 should ignore the dirname and use only the basename: markdown-1 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-markdown-1-default-1.com"
    >
    <title>
      Markdown 1
    </title>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`given the custom permalink page-1/ should ignore the dirname and use only the basename: page-1 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-page-1-default-1.com"
    >
    <title>
      Page 1
    </title>
  </head>
  <body>
    <h1>
      Page 1
    </h1>
  </body>
</html>
`;

exports[`given the path "/" should only match against the "/" policy path 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://single-slash.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h1>
      Home
    </h1>
  </body>
</html>
`;

exports[`given the path "/index.html" should only match against the "/" policy path 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://single-slash.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h1>
      Home
    </h1>
  </body>
</html>
`;

exports[`given the path "index.html" should only match against the "/" policy path 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://single-slash.com"
    >
    <title>
      Markdown Home
    </title>
  </head>
  <body>
    <h1>
      Home
    </h1>
  </body>
</html>
`;

exports[`when inline tag sources are enabled path: "/" should contain a hash+base64 source for each script/style, that merges with existing policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://frontmatter-home-default.com; script-src 'sha512-3j6cx4PFZZIQUupS7jwCN4QaghpVb95Q8mEsah38HTHT0BlCLn9pu8BDrMp0wT1kgpW2FE4H2cIXWE/dacTH0Q==' 'sha512-Mo2IittpCyzD+z7uC8fxA/AqYkcaSvAuqt0y2w0nIyWz+1lEQ5sKEEogY6vtxY+buvX+XPralLurIdDePvvokA=='; style-src 'sha512-xV3jXpd2QNEn/5AgkQKlit9GPjzfN2s5fFImBFKUe83y5/2KtaLxFvHBgvBEueD2DUnvwv1sCWofshvXJQGLmQ==' 'sha512-cbozKx9JFPkwg4AqdeM3KO0f5pA7gjKKGneGI9TFFOWxzPx+rsPdzihgqCd5UZFC2knhfiq/x2eD5gi5htkxlw=='"
    >
    <title>
      Markdown Home
    </title>
    <script>
      const a = 1;
    </script>
    <script>
      const b = 1;
const c = 1;
    </script>
    <style>
      div { color: red; }
    </style>
    <style>
      div {
  color: blue;
}
    </style>
  </head>
  <body>
    <h2 id="Home">
      Home
    </h2>
  </body>
</html>
`;

exports[`when inline tag sources are enabled path: "markdown-1" should contain a hash+base64 source for each script/style, that merges with existing policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://prod-markdown-1-default-1.com https://frontmatter-markdown-1-default-1.com; script-src 'sha512-3j6cx4PFZZIQUupS7jwCN4QaghpVb95Q8mEsah38HTHT0BlCLn9pu8BDrMp0wT1kgpW2FE4H2cIXWE/dacTH0Q==' 'sha512-Mo2IittpCyzD+z7uC8fxA/AqYkcaSvAuqt0y2w0nIyWz+1lEQ5sKEEogY6vtxY+buvX+XPralLurIdDePvvokA=='; style-src 'sha512-xV3jXpd2QNEn/5AgkQKlit9GPjzfN2s5fFImBFKUe83y5/2KtaLxFvHBgvBEueD2DUnvwv1sCWofshvXJQGLmQ==' 'sha512-cbozKx9JFPkwg4AqdeM3KO0f5pA7gjKKGneGI9TFFOWxzPx+rsPdzihgqCd5UZFC2knhfiq/x2eD5gi5htkxlw=='"
    >
    <title>
      Markdown 1
    </title>
    <script>
      const a = 1;
    </script>
    <script>
      const b = 1;
const c = 1;
    </script>
    <style>
      div { color: red; }
    </style>
    <style>
      div {
  color: blue;
}
    </style>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;

exports[`when inline tag sources are enabled path: "markdown-2" should contain a hash+base64 source for each script/style, that merges with existing policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="img-src https://frontmatter-markdown-2-img-1.com; script-src 'sha512-3j6cx4PFZZIQUupS7jwCN4QaghpVb95Q8mEsah38HTHT0BlCLn9pu8BDrMp0wT1kgpW2FE4H2cIXWE/dacTH0Q==' 'sha512-Mo2IittpCyzD+z7uC8fxA/AqYkcaSvAuqt0y2w0nIyWz+1lEQ5sKEEogY6vtxY+buvX+XPralLurIdDePvvokA=='; style-src 'sha512-xV3jXpd2QNEn/5AgkQKlit9GPjzfN2s5fFImBFKUe83y5/2KtaLxFvHBgvBEueD2DUnvwv1sCWofshvXJQGLmQ==' 'sha512-cbozKx9JFPkwg4AqdeM3KO0f5pA7gjKKGneGI9TFFOWxzPx+rsPdzihgqCd5UZFC2knhfiq/x2eD5gi5htkxlw=='"
    >
    <title>
      Markdown 2
    </title>
    <script>
      const a = 1;
    </script>
    <script>
      const b = 1;
const c = 1;
    </script>
    <style>
      div { color: red; }
    </style>
    <style>
      div {
  color: blue;
}
    </style>
  </head>
  <body>
    <h1>
      Markdown 1
    </h1>
  </body>
</html>
`;
